doc/manual/typing.md

```diff
@@ -0,0 +1,629 @@
+anonymous function, telling us it is expected to return a boolean instead of a
+number. Notice how we just had to give the top-level annotation `List Num`.
+Nickel performs type inference, so that you don't have to write the type for
+`filter`, the filtering function nor the list.
+
+### Take-away
+
+Nickel is gradually typed, meaning you can mix both static typing and dynamic
+typing. The default is dynamic typing. The static typechecker kicks in when
+using a type annotation `exp : Type`, which delimits a statically typed block.
+
+Nickel also has type inference, sparing you writing unnecessary type
+annotations.
+
+## Type system
+
+Let us now have a quick tour of the type system. The basic types are:
+
+- `Dyn`: the dynamic type. This is the type given to most expressions outside of
+  a typed block. A value of type `Dyn` can be pretty much anything.
+- `Num`: the only number type. Currently implemented as a 64bits float.
+- `Str`: a string.
+- `Bool`: a boolean, that is either `true` or `false`.
+<!-- - `Lbl`: a contract label. You usually don't need to use it or worry about it, -->
+<!--     it is more of an internal thing.  -->
+
+The following type constructors are available:
+
+- **List**: `List T`. A list of elements of type `T`. When no `T` is specified, `List`
+  alone is an alias for `List Dyn`.
+
+  Example:
+  ```nickel
+  let x : List (List Num) = [[1,2], [3,4]] in
+  lists.flatten x : List Num
+  ```
+- **Record**: `{field1: T1, .., fieldn: Tn}`. A record whose field
+  names are known statically as `field1`, .., `fieldn`, respectively of type
+  `T1`, .., `Tn`.
+
+  Example:
+  ```nickel
+  let pair : {fst: Num, snd: Str} = {fst = 1, snd = "a"} in
+  pair.fst : Num
+  ```
+- **Dynamic record**: `{_: T}`. A record whose field
+  names are statically unknown but are all of type `T`.  Typically used to model
+  dictionaries.
+
+  Example:
+  ```nickel
+  let occurences : {_: Num} = {a = 1, b = 3, c = 0} in
+  records.map (fun char count => count+1) occurences : {_ : Num}
+  ```
+- **Enum**: `<tag1, .., tagn>`: an enumeration comprised of alternatives `tag1`,
+  .., `tagn`. An enumeration literal is prefixed with a backtick and serialized
+  as a string. It is useful to encode finite alternatives. The advantage over
+  strings is that the typechecker handles them more finely: it is able to detect
+  incomplete matches, for example.
+
+  Example:
+  ```nickel
+  let protocol : <http, ftp, sftp> = `http in
+  (switch {
+    `http => 1,
+    `ftp => 2,
+    `sftp => 3
+  } protocol) : Num
+  ```
+- **Arrow (function)**: `S -> T`. A function taking arguments of type `S` and returning a value of
+  type `T`. For multi-parameters functions, just iterate the arrow constructor.
+
+  Example:
+  ```nickel
+  {
+    incr : Num -> Num = fun x => x + 1,
+    mkPath : Str -> Str -> Str -> Str = fun basepath filename ext =>
+      "#{basepath}/#{filename}.#{ext}",
+  }
+  ```
+
+### Polymorphism
+
+#### Plain polymorphism
```

Comment:
Maybe

```suggestion
#### Type polymorphism
```

I'm not sure this is good, but I do fine “plan polymorphism” fairly weird sounding.